pkg/chunk/aws_storage_client.go

```diff
@@ -209,40 +204,34 @@ func (a awsStorageClient) BatchWrite(ctx context.Context, input WriteBatch) erro
 		}
 
 		if err != nil {
-			for tableName := range reqs {
+			for tableName := range requests {
 				recordDynamoError(tableName, err, "DynamoDB.BatchWriteItem")
 			}
+
+			// If we get provisionedThroughputExceededException, then no items were processed,
+			// so back off and retry all.
+			if awsErr, ok := err.(awserr.Error); ok && ((awsErr.Code() == dynamodb.ErrCodeProvisionedThroughputExceededException) || request.Retryable()) {
+				unprocessed.TakeReqs(requests, -1)
+				backoff.backoff()
+				continue
+			}
+
+			// All other errors are critical.
+			return err
 		}
 
 		// If there are unprocessed items, backoff and retry those items.
 		if unprocessedItems := resp.UnprocessedItems; unprocessedItems != nil && dynamoDBWriteBatch(unprocessedItems).Len() > 0 {
 			unprocessed.TakeReqs(unprocessedItems, -1)
-			time.Sleep(backoff)
-			backoff = nextBackoff(backoff)
+			backoff.backoffWithoutCounting()
```

Comment:
Why does this not count the backoff? I see that the previous version also didn't increment the counter, but I can't recall why. If you know the answer, would you mind putting it in a comment? If not, feel free to disregard.

Reply:
No; I thought I did but reading [the docs](http://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_BatchWriteItem.html) dispels that notion.

The code comes from #256, without particular explanation.